Fixes #31692 - Send and receive katello-agent messages

jturel opened the pull request:
## Goals

- Preserve as much behavior as possible when issuing package actions on clients. Most of that related to what dynflow looks like during and after the task completes.
- Show katello_agent message handler status in the ping API
- Structure the code so that we don't loathe fixing bugs, debugging, refactoring if needed

## TODO

- [x] Fix tests
- [x] Rebase on https://github.com/Katello/katello/pull/9121 when merged (removes the EventDaemon rename from this PR)
- [x] Share qpid connection when dispatching the same client action to n > 1 hosts
- [x] Switch from qpid_messaging to qpid_proton https://github.com/Katello/katello/pull/9088/commits/b4f22b0bddd997574703ac48b45e74b672035398
- [x] Suspend actions instead of polling https://github.com/Katello/katello/pull/9088/commits/3be977bf715b92a6889ce9031be10bbcacb2c040
- [x] Remove erratum_install_batch_size setting https://github.com/Katello/katello/pull/9088/commits/a1dc10174bab48ca94cc8426d2c678b3832ca655

## The players
- `Actions::Katello::AgentAction`
  - superclass of all agent related dynflow actions
- `Actions::Katello::Agent::DispatchHistoryPresenter`
  - Shows result of the katello-agent action in the task details
- `Katello::Agent::Dispatcher`
  - service class responsible for issuing agent messages to content hosts
- `Katello::Agent::DispatchHistory`
  - Each dispatch action creates a history in the db which has a foreign key to the related host
- `Katello::EventDaemon::Services::AgentEventReceiver`
  - The event daemon service which connects to Qpid to receive client responses
- `Katello::Agent::ClientMessageHandler`
  - Once a client message is received, this code handles what to do - mostly updating the related `DispatchHistory`
- `Katello::Qpid::Connection`
  - Implementation of a qpid connection which uses qpid_proton gem
- `Katello::Agent::Connection`
  - Used by `Dispatcher` and `AgentEventReceiver` to connect to qpid consistently
- `Katello::Agent::BaseMessage`
  - All agent messages subclass `BaseMessage`. They set attributes and define methods that dictate what the message we send to clients look like


## Questionable choices (things you might especially have input on)

- Rather than sending an event to the dynflow action when we receive a message from the client to wake the task up, I used Dynflow's polling mechanism just like the current pulp-based actions do. It's not as elegant, but it avoids needing to know anything about Foreman Tasks from the client message handling side and I think that's an advantage. **Discussion concludes we will switch to sending events with the async approach**

- `Actions::Katello::AgentAction` is now the superclass of all katello-agent related Dynflow actions. This is different to the prior implementation where we planned a sub-action to handle issuing and polling the pulp task.

- Removed usage of `erratum_install_batch_size` which determined how many pulp tasks we spawned at once. Seems unnecessary now, so I can also add code to remove the setting. **Discussion concludes it's safe to remove this setting**

## Secondary outcomes
- Old pulp actions removed
- `Katello::EventDaemon` renamed to `Katello::EventDaemon::Runner` to make better use of `EventDaemon` namespace

## Testing
Highly recommended to spin up a second devel box for testing. It requires using Forklift to make sure these installer PRs are landed when the VM is set up: https://github.com/theforeman/puppet-katello/pull/377 https://github.com/theforeman/puppet-katello_devel/pull/249 https://github.com/theforeman/puppet-foreman_proxy_content/pull/322. `sudo yum install qpid-proton-cpp-devel` and bundle install after checking this PR out

Create the qpid queue:

```
qpid-config --ssl-certificate=/etc/pki/katello/qpid_router_client.crt --ssl-key=/etc/pki/katello/qpid_router_client.key -a amqps://localhost:5671 add queue katello.agent --durable
```

With the above in place, you can just  check this PR out, migrate the db, bundle install, and test host actions as usual while paying particular attention to consistent behavior between new and old.

ehelms commented on katello.gemspec:
```diff
@@ -39,6 +39,7 @@ Gem::Specification.new do |gem|
   gem.add_dependency "stomp"
   gem.add_dependency "scoped_search", ">= 4.1.9"
 
+  gem.add_dependency "qpid_messaging"
```
This is going to break on EL8 where we can't build this gem as an RPM currently. I think we should drop this requirement here and where you require the gem, do so conditionally if its present and otherwise don't and disable the functionality.

chris1984 commented:
@jturel what is a good way to test this on a current devel box?

jturel commented:
> @jturel what is a good way to test this on a current devel box?

Added testing notes. You should hold off for now as there are some changes coming based on the discussion yesterday. But you may wanna get a second vm prepped!

jturel commented on app/lib/katello/event_daemon/runner.rb:
```diff
@@ -0,0 +1,96 @@
+module Katello
```
This file was only renamed, and the Monitor class moved to its own file

jturel commented:
[test katello]

jturel commented:
[test katello]

jturel commented:
[test katello]

parthaa commented on app/services/katello/candlepin_event_listener.rb:
```diff
@@ -1,4 +1,5 @@
 module Katello
+  # TODO: Move this class to app/lib/katello/event_daemon/services with other service definitions
```
do you still intend to do this?

jturel replied:
I think it's out of scope of this PR so that's a TODO for some future dev (maybe me)

parthaa commented on app/services/katello/agent/dispatcher.rb:
```diff
@@ -0,0 +1,57 @@
+        fail("Couldn't find all hosts specified") unless host_ids.size == uuid_data.size
+
+        host_data = uuid_data.map do |host_id, consumer_id|
+          args[:consumer_id] = consumer_id
```
Is this intentional? would you want ```args[:consumer_id]``` to be the last host in the loop ?
Wouldn't be better to not have this line and say
``` ruby
message: message_class.new(**(args.merge(consumer_id: consumer_id)))
```

jturel replied:
I like your idea! Better to not modify the params even though it wasn't a problem here (afaik)

parthaa commented on app/lib/katello/messaging/received_message.rb:
```diff
@@ -1,12 +1,26 @@
       end
+
+      def body
+        return @json_body if @json_body
```
What is the  point of this? Who is setting it up?

parthaa commented on app/lib/katello/agent/client_message_handler.rb:
```diff
@@ -0,0 +1,63 @@
+      end
+
+      # rubocop:disable Metrics/MethodLength
+      def self.handle(message)
```
can you break this up to smaller methods.

jturel replied:
Let me think about that. I'm all for small methods but I'm not sure those two suggested methods serve their own individual purpose

parthaa commented on app/lib/actions/katello/host/package/update.rb:
```diff
@@ -2,18 +2,13 @@ module Actions
-                        args:          packages)
-            plan_self(:host_id => host.id)
+          def agent_action_type
+            :content_install
```
should this be ```:content_update```

jturel replied:
I don't think so. See the usage in DispatchHistoryPresenter and compare to behavior in the current presenter

parthaa commented on app/lib/actions/katello/agent_action.rb:
```diff
@@ -0,0 +1,106 @@
+      end
+
+      def agent_action_type
+        nil
```
so nil agent action type is ok?

jturel replied:
Yeah, see how it's used in DispatchHistoryPresenter

parthaa commented on app/lib/actions/katello/agent_action.rb:
```diff
@@ -0,0 +1,106 @@
+        action_subject(host)
+
+        # if already dispatched by bulk action use the provided history ID
+        dispatch_history_id = options[:dispatch_histories]&.dig(host.id.to_s)
```
will this do the same thing ?

```suggestion
        dispatch_history_id = options.dig(:dispatch_histories, host.id.to_s)
```

jturel replied:
Yes!

parthaa commented:
Code wise looks mostly good to me except for a couple of suggestions and questions. 
Going to run and test.

ehelms commented:
I am seeing:

```
2021-02-03T03:46:42 [I|bac|8f5e14c3] Task {label: Actions::Katello::Host::Erratum::Install, id: bca34efd-8b17-4c36-848d-fd442a701d83, execution_plan_id: 8d8e584b-6403-46a2-b379-a68ae7b2f872} state changed: planning
2021-02-03T03:46:42 [E|bac|] no implicit conversion of Symbol into Integer (TypeError)
 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/lib/actions/katello/agent_action.rb:19:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:514:in `block (3 levels) in execute_plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:36:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:36:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_request_id.rb:10:in `block in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_request_id.rb:31:in `with_current_request_id'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_request_id.rb:9:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_timezone.rb:10:in `block in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_timezone.rb:31:in `with_current_timezone'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_timezone.rb:9:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_user.rb:10:in `block in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_user.rb:31:in `with_current_user'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_user.rb:9:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_taxonomies.rb:10:in `block in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_taxonomies.rb:30:in `with_current_taxonomies'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-3.0.3/app/lib/actions/middleware/keep_current_taxonomies.rb:9:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:36:in `plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/world.rb:31:in `execute'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:513:in `block (2 levels) in execute_plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan.rb:384:in `switch_flow'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:417:in `concurrence'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:512:in `block in execute_plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:472:in `block in with_error_handling'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:472:in `catch'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:472:in `with_error_handling'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:511:in `execute_plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/action.rb:285:in `execute'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan/steps/plan_step.rb:55:in `block in execute'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan/steps/abstract.rb:167:in `with_meta_calculation'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan/steps/plan_step.rb:54:in `execute'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan.rb:286:in `block (2 levels) in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan.rb:384:in `switch_flow'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan.rb:374:in `with_planning_scope'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/execution_plan.rb:285:in `block in plan'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:19:in `pass'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware.rb:44:in `plan_phase'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:23:in `call'
 | /opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.7/lib/dynflow/middleware/stack.rb:27:in `pass'
```

ehelms commented:
I cannot tell if this is related or not:

```
2021-02-03T15:10:42 [E|kat|f8ee9620] Error handling Candlepin event
2021-02-03T15:10:42 [E|kat|f8ee9620] no implicit conversion of ActiveSupport::HashWithIndifferentAccess into String
2021-02-03T15:10:42 [E|kat|f8ee9620] /opt/rh/rh-ruby25/root/usr/share/ruby/json/common.rb:156:in `initialize'
 f8ee9620 | /opt/rh/rh-ruby25/root/usr/share/ruby/json/common.rb:156:in `new'
 f8ee9620 | /opt/rh/rh-ruby25/root/usr/share/ruby/json/common.rb:156:in `parse'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin/message_handler.rb:16:in `content'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin/message_handler.rb:20:in `event_data'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin/message_handler.rb:28:in `status'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin/event_handler.rb:45:in `reindex_subscription_status'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin/event_handler.rb:30:in `handle'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin_event_listener.rb:59:in `block in handle_message'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/lib/katello/util/support.rb:76:in `with_db_connection'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin_event_listener.rb:56:in `handle_message'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/services/katello/candlepin_event_listener.rb:39:in `block in run'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/katello-4.0.0.pre.master/app/lib/katello/messaging/stomp_connection.rb:45:in `block in subscribe'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/stomp/client.rb:189:in `block in subscribe'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/client/utils.rb:154:in `find_listener'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/client/utils.rb:184:in `block in create_listener_maps'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/client/utils.rb:203:in `block (2 levels) in start_listeners'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/client/utils.rb:194:in `loop'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/stomp-1.4.9/lib/client/utils.rb:194:in `block in start_listeners'
 f8ee9620 | /opt/theforeman/tfm/root/usr/share/gems/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

jturel commented:
@partha I think I've addressed your comments. To test you'll need to apply this PR that came in today to your env: https://github.com/theforeman/puppet-foreman_proxy_content/pull/322

Shortcut for it is to add this in /etc/qpid-dispatch/qdrouterd.conf:

```
linkRoute {
    prefix: katello.agent
    dir: in
    connection: broker
}
```

ehelms reviewed (approved):
This is ready to go from my testing

parthaa commented on app/lib/katello/event_daemon/services/agent_event_receiver.rb:
```diff
@@ -14,7 +14,7 @@ def initialize
           def handle(message)
             ::Katello::Util::Support.with_db_connection do
-              ::Katello::Agent::ClientMessageHandler.handle(message)
+              ::Katello::Agent::ClientMessageHandler.new(message).handle
```
Love it. Was thinking along the same -> https://refactoring.com/catalog/replaceFunctionWithCommand.html

parthaa commented:
I seem to be sporadically hitting issue with this PR. Filed https://projects.theforeman.org/issues/31814 . Recommend merging this and revisiiting.
